Remove non-AsXx surface area from M.E.AI.OpenAI/AzureAIInference

stephentoub opened the pull request:
We expect the AsChatClient/AsEmbeddingGenerator extension methods from M.E.AI.OpenAI and M.E.AI.AzureAIInference to move into the OpenAI / Azure.AI.Inference libraries, respectively. To prepare for that, and the temporary M.E.AI.OpenAI/AzureAIInference libs then being deprecated, this PR removes the other surface area from these assemblies. For anything folks found useful, we should find another way to ship it, likely just as sample source somewhere.

Closes https://github.com/dotnet/extensions/issues/6094
Closes https://github.com/dotnet/extensions/issues/5965
Closes https://github.com/dotnet/extensions/issues/5964

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/extensions/pull/6138)

eiriktsarpalis commented on src/Libraries/Microsoft.Extensions.AI.AzureAIInference/AzureAIInferenceChatClient.cs:
```diff
@@ -321,7 +311,7 @@ private ChatCompletionsOptions ToAzureAIOptions(IEnumerable<ChatMessage> chatCon
                             if (prop.Value is not null)
                             {
-                                byte[] data = JsonSerializer.SerializeToUtf8Bytes(prop.Value, ToolCallJsonSerializerOptions.GetTypeInfo(typeof(object)));
+                                byte[] data = JsonSerializer.SerializeToUtf8Bytes(prop.Value, AIJsonUtilities.DefaultOptions.GetTypeInfo(typeof(object)));
```
I'd forgotten about `AdditionalProperties`. Hardcoding to the default options effectively forces known types only in AOT.

stephentoub replied:
I suspect primitives is sufficient. We can always add an overload later that takes a JSO if it proves important. Or you think it should be done now?

eiriktsarpalis replied:
I suspect there always will be a need for storing something less trivial than just primitives.

stephentoub replied:
Quite possibly, but this in particular is about properties that should be sent up to the service as extra properties that a given service understands:
https://learn.microsoft.com/en-us/rest/api/aifoundry/modelinference/#extensibility
and those are generally going to be primitives, like float, int, or string.

(Also, the Azure.AI.Inference library itself is not yet AOT friendly.)

eiriktsarpalis replied:
What about `float[]`, `int?`, `int?[]`, or `string[]`? If we're drawing a line in the sand we probably need to document it (and quite possibly reject values whose type is not supported at runtime). Otherwise we find ourselves in a situation where clients that work without issue in CoreCLR end up failing in Native AOT without a workaround.

stephentoub replied:
? What about float[], int?, int?[], or string[]?

`int?` boxes the same as `int`, but regardless it's already represented in the default options.

If we think `float[]`, `int?[]`, or `string[]` will be desirable, we can just add them to AIJsonUtilities.

SteveSandersonMS commented:
This is certainly all good stuff in general.

Within the `GetService` implementations, should we also remove the clauses that return `this` if  `serviceType.IsInstanceOfType(this)`? I'm not sure what use case those clauses would serve now you can't just reference the type directly and can only get it in obscure ways (e.g., by being the second-to-last middleware and passing `innerClient.GetType()`.

SteveSandersonMS commented on src/Libraries/Microsoft.Extensions.AI.OpenAI/OpenAIRealtimeExtensions.cs:
```diff
@@ -1,140 +0,0 @@
-/// <summary>
-/// Provides extension methods for working with <see cref="RealtimeConversationSession"/> and related types.
-/// </summary>
-public static class OpenAIRealtimeExtensions
```
Losing this is a bit unfortunate (for me anyway, I'm certainly using it from various projects and from https://github.com/SteveSandersonMS/dotnet-ai-workshop). 

Isn't there's still legitimacy in the provider-specific packages including helpers for making more of the provider-specific functionality work with MEAI, even if it's not purely an `IChatClient`?

stephentoub commented:
This should be ready to go.

One important note:
I've made two additional changes:
- Removed the AsChatClient/AsEmbeddingGenerator extensions for the OpenAIClient type, leaving just the ones for OpenAI.Chat.ChatClient and OpenAI.Embeddings.EmbeddingClient. Given what we've seen with OpenAI's move to push folks away from ChatClient and towards ResponseClient, it seems unwise for us to have AsChatClient on OpenAIClient that's bound to the chat completion implementation.
- Renamed AsChatClient to AsIChatClient and AsEmbeddingGenerator to AsIEmbeddingGenerator. This is to avoid confusion with naming, e.g. having an OpenAI `ChatClient` have an `AsChatClient` extension method on it.

cc: @jeffhandley, @SteveSandersonMS

shyamnamboodiripad reviewed (approved):
Approving for the single change in the eval test 👍🏾

SteveSandersonMS commented on test/Libraries/Microsoft.Extensions.AI.OpenAI.Tests/OpenAIResponseClientTests.cs:
```diff
@@ -134,7 +129,6 @@ public async Task BasicRequestResponse_NonStreaming()
               "tool_choice": "auto",
               "tools": [],
               "top_p": 1.0,
-              "truncation": "auto",
```
Out of interest, why remove this? Is it something OpenAI has changed?

stephentoub replied:
Yes, it was removed with the latest rev of responses.

stephentoub commented:
@SteveSandersonMS, @jeffhandley, per my discussion with Steve yesterday, I temporarily added back the AsChatClient(OpenAIClient) and AsEmbeddingGenerator(OpenAIClient) extensions, but as EBNever and Obsolete.
